native/dgebak.go

```diff
@@ -0,0 +1,79 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dgebak transforms an n×m matrix V as
+//  V = P D V,        if side == blas.Right,
+//  V = P D^{-1} V,   if side == blas.Left,
+// where P and D are n×n permutation and scaling matrices, respectively,
+// implicitly represented by job, scale, ilo and ihi as returned by Dgebal.
+//
+// Typically, columns of the matrix V contain the right or left (determined by
+// side) eigenvectors of the balanced matrix output by Dgebal, and Dgebak forms
+// the eigenvectors of the original matrix.
+//
+// Dgebak is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgebak(job lapack.Job, side blas.Side, n, ilo, ihi int, scale []float64, m int, v []float64, ldv int) {
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.None, lapack.Permute, lapack.Scale, lapack.PermuteScale:
+	}
+	switch side {
+	default:
+		panic(badSide)
+	case blas.Left, blas.Right:
+	}
+	checkMatrix(n, m, v, ldv)
+	switch {
+	case ilo < 0 || max(0, n-1) < ilo:
+		panic(badIlo)
+	case ihi < min(ilo, n-1) || n <= ihi:
+		panic(badIhi)
+	}
+
+	// Quick return if possible.
+	if n == 0 || m == 0 || job == lapack.None {
+		return
+	}
+
+	bi := blas64.Implementation()
+	if ilo != ihi && job != lapack.Permute {
+		// Backward balance.
+		if side == blas.Right {
+			for i := ilo; i <= ihi; i++ {
+				bi.Dscal(m, scale[i], v[i*ldv:], 1)
+			}
+		} else {
+			for i := ilo; i <= ihi; i++ {
+				bi.Dscal(m, 1/scale[i], v[i*ldv:], 1)
+			}
+		}
+	}
+	if job == lapack.Scale {
+		return
+	}
+	// Backward permutation.
+	for i := ilo - 1; i >= 0; i-- {
```

Comment:
I don't know why Lapack distinguishes between sides. The relations in the comment show it and the lapack code confirms it that the permutations are done identically for both sides.